### ISSUE TITLE
Can't evaluate criteria against alternate class <class 'app.sec_models.MyUser'>

### ISSUE THREAD

**Issue by soundmaking** [dpgaspar/Flask-AppBuilder#1822]
Not sure exactly when the issue began, but it might have been when upgrading from `Flask-AppBuilder==3.1.1` to `Flask-AppBuilder==3.4.4`

### Environment

Flask-Appbuilder version: 3.4.4

pip freeze output:

alembic==1.5.2
apispec==3.3.2
attrs==20.3.0
Babel==2.9.0
black==22.1.0
blinker==1.4
cachetools==4.2.1
certifi==2020.12.5
cffi==1.14.4
chardet==4.0.0
click==8.0.4
colorama==0.4.4
defusedxml==0.6.0
dnspython==2.1.0
email-validator==1.1.2
Flask==1.1.2
Flask-AppBuilder==3.4.4
Flask-Babel==1.0.0
Flask-JWT-Extended==3.25.0
Flask-Login==0.4.1
Flask-Mail==0.9.1
flask-marshmallow==0.11.0
Flask-Migrate==2.6.0
Flask-Misaka==1.0.0
Flask-OpenID==1.2.5
Flask-SQLAlchemy==2.4.4
Flask-WTF==0.14.3
google-api-core==1.25.0
google-api-python-client==1.12.8
google-auth==1.24.0
google-auth-httplib2==0.0.4
google-auth-oauthlib==0.4.2
googleapis-common-protos==1.52.0
gspread==3.6.0
httplib2==0.18.1
idna==2.10
importlib-metadata==3.4.0
itsdangerous==1.1.0
Jinja2==2.11.2
jsonschema==3.2.0
Mako==1.1.4
MarkupSafe==1.1.1
marshmallow==3.10.0
marshmallow-enum==1.5.1
marshmallow-sqlalchemy==0.23.1
misaka==2.1.1
mypy-extensions==0.4.3
oauth2client==4.1.3
oauthlib==3.1.0
pathspec==0.9.0
pkg_resources==0.0.0
platformdirs==2.5.1
prison==0.2.1
protobuf==3.14.0
pyasn1==0.4.8
pyasn1-modules==0.2.8
pycparser==2.20
PyJWT==1.7.1
pyrsistent==0.17.3
python-dateutil==2.8.1
python-editor==1.0.4
python3-openid==3.2.0
pytz==2020.5
PyYAML==5.4.1
requests==2.25.1
requests-oauthlib==1.3.0
rsa==4.7
sdfspu==0.2.1
sheetfu==1.5.4
six==1.15.0
SQLAlchemy==1.3.22
SQLAlchemy-Utils==0.36.8
tomli==2.0.1
typed-ast==1.5.2
typing_extensions==4.1.1
uritemplate==3.0.1
urllib3==1.26.2
Werkzeug==1.0.1
WTForms==2.3.3
zipp==3.4.0

### Describe the expected results

Expected no errors on /login.

### Describe the actual results

I was checking for something else in the error.log on app in production, and instead found many error 
with `raise exception
sqlalchemy.exc.InvalidRequestError: Could not evaluate current criteria in Python: "Can't evaluate criteria against alternate class <class 'app.sec_models.MyUser'>". Specify 'fetch' or False for the synchronize_session execution option.
`

From the error.log:

```pytb
2022-03-24 21:30:43,564: Exception on /login/ [POST]
Traceback (most recent call last):
  File "/home/client/envs/ppoc2022/lib/python3.7/site-packages/sqlalchemy/orm/persistence.py", line 1938, in _do_pre_synchronize_evaluate
    eval_condition = evaluator_compiler.process(*crit)
  File "/home/client/envs/ppoc2022/lib/python3.7/site-packages/sqlalchemy/orm/evaluator.py", line 85, in process
    return meth(clause)
  File "/home/client/envs/ppoc2022/lib/python3.7/site-packages/sqlalchemy/orm/evaluator.py", line 181, in visit_binary
    map(self.process, [clause.left, clause.right])
  File "/home/client/envs/ppoc2022/lib/python3.7/site-packages/sqlalchemy/orm/evaluator.py", line 85, in process
    return meth(clause)
  File "/home/client/envs/ppoc2022/lib/python3.7/site-packages/sqlalchemy/orm/evaluator.py", line 107, in visit_column
    % parentmapper.class_
sqlalchemy.orm.evaluator.UnevaluatableError: Can't evaluate criteria against alternate class <class 'app.sec_models.MyUser'>
**NO MATCH**
The above exception was the direct cause of the following exception:
**NO MATCH**
Traceback (most recent call last):
  File "/home/client/envs/ppoc2022/lib/python3.7/site-packages/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/client/envs/ppoc2022/lib/python3.7/site-packages/flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/client/envs/ppoc2022/lib/python3.7/site-packages/flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/home/client/envs/ppoc2022/lib/python3.7/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/home/client/envs/ppoc2022/lib/python3.7/site-packages/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/client/envs/ppoc2022/lib/python3.7/site-packages/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/home/client/envs/ppoc2022/lib/python3.7/site-packages/flask_appbuilder/security/views.py", line 518, in login
    form.username.data, form.password.data
  File "/home/client/envs/ppoc2022/lib/python3.7/site-packages/flask_appbuilder/security/manager.py", line 874, in auth_user_db
    self.noop_user_update(first_user)
  File "/home/client/envs/ppoc2022/lib/python3.7/site-packages/flask_appbuilder/security/sqla/manager.py", line 247, in noop_user_update
    self.get_session.execute(stmt)
  File "<string>", line 2, in execute
  File "/home/client/envs/ppoc2022/lib/python3.7/site-packages/sqlalchemy/orm/session.py", line 1646, in execute
    _parent_execute_state is not None,
  File "/home/client/envs/ppoc2022/lib/python3.7/site-packages/sqlalchemy/orm/persistence.py", line 1823, in orm_pre_session_exec
    update_options,
  File "/home/client/envs/ppoc2022/lib/python3.7/site-packages/sqlalchemy/orm/persistence.py", line 1951, in _do_pre_synchronize_evaluate
    from_=err,
  File "/home/client/envs/ppoc2022/lib/python3.7/site-packages/sqlalchemy/util/compat.py", line 207, in raise_
    raise exception
sqlalchemy.exc.InvalidRequestError: Could not evaluate current criteria in Python: "Can't evaluate criteria against alternate class <class 'app.sec_models.MyUser'>". Specify 'fetch' or False for the synchronize_session execution option.

```

### From app/sec_models.py
```python
"""Models needed by the security manager
"""

import datetime

from flask_appbuilder import Model
from flask_appbuilder.models.mixins import AuditMixin
from flask_appbuilder.security.sqla.models import User
from sqlalchemy import Column, Integer, ForeignKey, String, Table, DateTime, Text
from sqlalchemy.orm import relationship

from .util_models import AuditMixinExtra


class MyUser(User):
    __tablename__ = 'ab_user'
    extra = Column(String(256))

    def __repr__(self):
        return str(self.username)


assoc_user_emailaddr = Table(
    'user_emailaddr',
    Model.metadata,
    Column('id', Integer, primary_key=True),
    Column('user_id', Integer, ForeignKey('ab_user.id')),
    Column('email_addr', Integer, ForeignKey('email_address_record.id')),
)


class EmailAddressRecord(AuditMixinExtra, Model):
    """An EmailAddress might be seen before a user is registered with it;

    this object has backref from other objects, such as GoogleFormResponse,
    which include an email address.

    If a user then does register, it can link to other objects via this record.
    """

    id = Column(Integer, primary_key=True)
    email = Column(String(64), unique=True, nullable=False)
    user = relationship(
        'MyUser',
        secondary=assoc_user_emailaddr,
        backref='email_address_record'
    )

    def __repr__(self):
        return '@:{}'.format(self.id)

```

### Steps to reproduce

I am still investigating the situation, but posting this incomplete report now with the hopes that someone might have enough info in the above to suggest what might be the issue.  

I intend to to post again when I have have reproduced the error.

Thanks for reading.

**Comment by dpgaspar** [comment]
Interesting, seems like it comes from: https://github.com/dpgaspar/Flask-AppBuilder/blob/e2d3f017bfeb30e863a25fa63de4699db20bc8d4/flask_appbuilder/security/sqla/manager.py#L241

please update if you reach any conclusion

**Comment by soundmaking** [comment]
Thanks, Daniel.  Yes, I will look at how my MyUser class, or other custom security manager stuff might be interacting with that no-op function...

### further detail

#### Login values tested to reproduce the error

- non-valid username: ERROR as described above, user get a server error 500 page
- valid username, valid password: OK, no error
- valid username, wrong password: OK, no error, FAB gives flash message as expected


#### from config.py
```python
# ...
SQLALCHEMY_DATABASE_URI = 'sqlite:///' + os.path.join(basedir, 'appdata.db') 
# ...
AUTH_TYPE = AUTH_DB
# ...
AUTH_TYPE = AUTH_DB
```

**Comment by soundmaking** [comment]
This seems to have fixed it: letting the noop update function use `self.user_model` instead of `User` on this line

https://github.com/dpgaspar/Flask-AppBuilder/blob/e2d3f017bfeb30e863a25fa63de4699db20bc8d4/flask_appbuilder/security/sqla/manager.py#L243


Added in my app/sec.py:

```python
from sqlalchemy import update
# . . . 

class MySecurityManager(SecurityManager):
    # . . .
    # minor change to SecurityManager.noop_user_update
    def noop_user_update(self, user: "User") -> None:
        stmt = (
            update(self.user_model)  # was update(User)
                .where(self.user_model.id == user.id)
                .values(login_count=user.login_count)
        )
        self.get_session.execute(stmt)
        self.get_session.commit()
```

**Comment by dpgaspar** [comment]
Right makes perfect sense @soundmaking, I'll make a PR for the fix

**Comment by dpgaspar** [comment]
@soundmaking ok merged the fix, will be out on the next release 4.0.1 (should be soon 1/2 weeks)
